### ISSUE TITLE
How to actually compile anything?

### ISSUE THREAD

**Issue by achingbrain** [matthew-james/esp-vagrant#3]
I've run `vagrant up`, `vagrant ssh` and `esp-open-sdk install`.  Everything completed without error.  Now I try to compile the examples:

```
[vagrant@vagrant-ubuntu-trusty-64 IoT_Demo]$ pwd
/home/vagrant/esp-open-sdk/sdk/examples/IoT_Demo
[vagrant@vagrant-ubuntu-trusty-64 IoT_Demo]$ make
make: Nothing to be done for `FORCE'.
```

I read somewhere that you should copy the project you want to compile into the `sdk` directory:

```
[vagrant@vagrant-ubuntu-trusty-64 sdk]$ pwd
/home/vagrant/esp-open-sdk/sdk
[vagrant@vagrant-ubuntu-trusty-64 sdk]$ cp -r examples/IoT_Demo .
[vagrant@vagrant-ubuntu-trusty-64 sdk]$ cd IoT_Demo/
[vagrant@vagrant-ubuntu-trusty-64 IoT_Demo]$ make
../../Makefile:200: warning: overriding commands for target `clean'
../Makefile:258: warning: ignoring old commands for target `clean'
You cloned without --recursive, fetching submodules for you.
git submodule update --init --recursive
make -C crosstool-NG -f ../Makefile _ct-ng
make: *** crosstool-NG: No such file or directory.  Stop.
make: *** [crosstool-NG/ct-ng] Error 2
```

The `crosstool-NG`, `lx106-hal` and `esptool` directories that would be created as git submodules are there so the error sounds misleading:

```
[vagrant@vagrant-ubuntu-trusty-64 esp-open-sdk]$ pwd
/home/vagrant/esp-open-sdk
[vagrant@vagrant-ubuntu-trusty-64 esp-open-sdk]$ ls -la
total 8748
drwxrwxr-x  8 vagrant vagrant    4096 Apr 26 23:00 .
drwxr-xr-x  6 vagrant vagrant    4096 Apr 26 18:06 ..
-rw-rw-r--  1 vagrant vagrant     199 Apr 26 18:06 crosstool-config-overrides
drwxrwxr-x 18 vagrant vagrant    4096 Apr 26 18:07 crosstool-NG
-rw-rw-r--  1 vagrant vagrant    1101 Apr 26 18:06 c_types-c99.patch
drwxrwxr-x 11 vagrant vagrant    4096 Apr 27 06:29 esp_iot_sdk_v1.0.1
-rw-rw-r--  1 vagrant vagrant 8880579 Apr 24 14:44 esp_iot_sdk_v1.0.1_15_04_24.zip
drwxrwxr-x  2 vagrant vagrant    4096 Apr 26 18:06 esptool
drwxrwxr-x  9 vagrant vagrant    4096 Apr 26 18:06 .git
-rw-rw-r--  1 vagrant vagrant     288 Apr 26 18:06 .gitmodules
drwxrwxr-x  6 vagrant vagrant    4096 Apr 26 18:46 lx106-hal
-rw-rw-r--  1 vagrant vagrant    7090 Apr 26 18:06 Makefile
-rw-rw-r--  1 vagrant vagrant    4469 Apr 26 18:06 README.md
-rw-rw-r--  1 vagrant vagrant   12265 Apr 24 22:19 release_note.txt
lrwxrwxrwx  1 vagrant vagrant      18 Apr 26 23:00 sdk -> esp_iot_sdk_v1.0.1
-rw-rw-r--  1 vagrant vagrant       0 Apr 26 18:46 .sdk_patch_1.0.1
drwxrwxr-x  8 vagrant vagrant    4096 Apr 26 18:43 xtensa-lx106-elf
```

The toolchain is on the PATH:

```
[vagrant@vagrant-ubuntu-trusty-64 esp-open-sdk]$ echo $PATH
/home/vagrant/esp-open-sdk/xtensa-lx106-elf/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
```

Any ideas?

**Comment by matthew-james** [comment]
Hello,

I haven't used the IoT demo yet.  I have been using the source-code-examples repo's makefile to compile everything.  I wrote a quick example here:

https://github.com/matthew-james/esp-vagrant/wiki/burning-firmware

Let me know if that works for you.

**Comment by matthew-james** [comment]
I think copying the project to the ````/home/vagrant/esp-open-sdk```` instead of ````/home/vagrant/esp-open-sdk/sdk```` would fix the error, as it's trying to cd into ````crosstool-NG```` when the error occurs.

You can also check the Makefile with ````make --just-print```` to see what is happening.  

I will check tonight and see if I can get it to work.

**Comment by achingbrain** [comment]
After copying the IoT_Demo directory to `/home/vagrant/esp-open-sdk` gives me:

```
[vagrant@vagrant-ubuntu-trusty-64 IoT_Demo]$ pwd
/home/vagrant/esp-open-sdk/IoT_Demo
[vagrant@vagrant-ubuntu-trusty-64 IoT_Demo]$ make
You cloned without --recursive, fetching submodules for you.
git submodule update --init --recursive
make -C crosstool-NG -f ../Makefile _ct-ng
make: *** crosstool-NG: No such file or directory.  Stop.
make: *** [crosstool-NG/ct-ng] Error 2
```

`make --just-print` gives me:

```
[vagrant@vagrant-ubuntu-trusty-64 IoT_Demo]$ make --just-print
echo "You cloned without --recursive, fetching submodules for you."
git submodule update --init --recursive
make -C crosstool-NG -f ../Makefile _ct-ng
make -C crosstool-NG -f ../Makefile _toolchain
cp esptool/esptool.py /home/vagrant/esp-open-sdk/IoT_Demo/xtensa-lx106-elf/bin/
make: *** No rule to make target `/home/vagrant/esp-open-sdk/IoT_Demo/xtensa-lx106-elf/xtensa-lx106-elf/sysroot/lib/libc.a', needed by `/home/vagrant/esp-open-sdk/IoT_Demo/xtensa-lx106-elf/xtensa-lx106-elf/sysroot/lib/libcirom.a'.  Stop.
```

Cloning [https://github.com/matthew-james/source-code-examples.git](https://github.com/matthew-james/source-code-examples.git) and running `make` in the `blinky` seems to work so it's not all bad:

```
[vagrant@vagrant-ubuntu-trusty-64 blinky]$ pwd
/home/vagrant/source-code-examples/blinky
[vagrant@vagrant-ubuntu-trusty-64 blinky]$ make
CC user/user_main.c
AR build/app_app.a
LD build/app.out
FW firmware/
```

**Comment by matthew-james** [comment]
So,

It looks like the "Recursion Magic" section is causing make to check every parent directory and run *that* makefile, so that error is actually coming from the esp-open-sdk makefile.

There isn't a default target in the makefile according to [this issue](https://github.com/pfalcon/esp-open-sdk/issues/10).

[Here is a Makefile that compiles correctly](https://gist.github.com/matthew-james/ce04dc0bf8a489ee3487).

Basically what I did was take the Makefile from the source-code-examples and add the needed  ````LIBS````.

I copied the IoT_Demo to user_data/IoT_Demo, then replaced the makefile with the new one.  After that it compiled correctly.

I am leaving this issue open until the wiki is updated.

**Comment by matthew-james** [comment]
I updated the wiki with guides on compiling the Espressif demos, NodeMCU firmware, and the source-code-examples [here](https://github.com/matthew-james/esp-vagrant/wiki/burning-firmware).  Proper makefiles are linked for all of the Espressif demos.

Thanks,
Matt

**Comment by achingbrain** [comment]
That's awesome thanks.  Makefiles are a dark art, one I'm not versed in.
